storage/s3.go

```diff
@@ -1274,3 +1356,60 @@ func generateRetryID() *string {
 	num, _ := rand.Int(rand.Reader, big.NewInt(math.MaxInt64))
 	return aws.String(num.String())
 }
+
+// EventStreamDecoder decodes a s3.Event with
+// the given decoder.
+type EventStreamDecoder interface {
+	Decode() ([]byte, error)
+}
+
+type JSONDecoder struct {
+	decoder *json.Decoder
+}
+
+func NewJSONDecoder(reader io.Reader) EventStreamDecoder {
+	return &JSONDecoder{
+		decoder: json.NewDecoder(reader),
+	}
+}
+
+func (jd *JSONDecoder) Decode() ([]byte, error) {
+	var val json.RawMessage
+	err := jd.decoder.Decode(&val)
+	if err != nil {
+		return nil, err
+	}
+	return val, nil
+}
+
+type CsvDecoder struct {
+	decoder   *csv.Reader
+	delimiter string
+}
+
+func NewCsvDecoder(reader io.Reader) EventStreamDecoder {
+	csvDecoder := &CsvDecoder{
+		decoder:   csv.NewReader(reader),
+		delimiter: ",",
+	}
+	// returned values from AWS has double quotes in it
+	// so we enable lazy quotes
+	csvDecoder.decoder.LazyQuotes = true
+	return csvDecoder
+}
+
+func (cd *CsvDecoder) Decode() ([]byte, error) {
+	res, err := cd.decoder.Read()
+	if err != nil {
+		return nil, err
+	}
+
+	result := []byte{}
+	for i, str := range res {
+		if i != len(res)-1 {
+			str = fmt.Sprintf("%s%s", str, cd.delimiter)
+		}
+		result = append(result, []byte(str)...)
+	}
```

Comment:
CSV decoder decodes fields to a `[]string`.  We cast those strings to bytes by populating each word with the delimiter and write to `result` slice, so we can pass the result directly to the channel without changing the signature:

https://github.com/denizsurmeli/s5cmd/blob/242617116034a62f48cfed3ca8865a63196e0eea/storage/s3.go#L664

 `byte[]` seemed more feasible since the `Write` to `stdout` also takes `[]byte`.

Reply:
👍